pybricks/media/pb_type_image.c

```diff
@@ -0,0 +1,353 @@
+    // memory allocator and use a python object to handle memory release.
+    uint8_t *buffer;  // m_malloc'd buffer, may be shared, NULL for display.
+    pbio_image_t image;
+    bool is_display;
+} pb_type_Image_obj_t;
+
+static int get_color(mp_obj_t obj) {
+    uint8_t max = pbdrv_display_get_max_value();
+    if (obj == mp_const_none) {
+        return max;
+    }
+    if (mp_obj_is_int(obj)) {
+        return mp_obj_get_int(obj);
+    }
+    const pbio_color_hsv_t *hsv = pb_type_Color_get_hsv(obj);
+    int32_t v = pbio_int_math_bind(hsv->v, 0, 100);
+    return max - v * max / 100;
+}
+
+mp_obj_t pb_type_Image_display_obj_new(void) {
+    pb_type_Image_obj_t *self = mp_obj_malloc(pb_type_Image_obj_t, &pb_type_Image);
+    self->buffer = NULL;
+    self->is_display = true;
+    self->image = *pbdrv_display_get_image();
+
+    return MP_OBJ_FROM_PTR(self);
+}
+
+static mp_obj_t pb_type_Image_make_new(const mp_obj_type_t *type,
+    size_t n_args, size_t n_kw, const mp_obj_t *args) {
+    PB_PARSE_ARGS_CLASS(n_args, n_kw, args,
+        PB_ARG_REQUIRED(source),
+        PB_ARG_DEFAULT_FALSE(sub),
+        PB_ARG_DEFAULT_INT(x1, 0),
+        PB_ARG_DEFAULT_INT(y1, 0),
+        PB_ARG_DEFAULT_NONE(x2),
+        PB_ARG_DEFAULT_NONE(y2));
+
+    pb_type_Image_obj_t *self = mp_obj_malloc(pb_type_Image_obj_t, type);
+
+    if (mp_obj_is_type(source_in, &pb_type_Image)) {
+        pb_type_Image_obj_t *source = MP_OBJ_TO_PTR(source_in);
+        if (!mp_obj_is_true(sub_in)) {
+            // Copy.
+            int width = source->image.width;
+            int height = source->image.height;
+            // TODO: avoid using MicroPython heap for large allocation.
+            self->buffer = m_malloc(width * height * sizeof(uint8_t));
+            self->is_display = false;
+            pbio_image_init(&self->image, self->buffer, width, height, width);
+            pbio_image_draw_image(&self->image, &source->image, 0, 0);
+        } else {
+            // Sub-image.
+            mp_int_t x1 = pb_obj_get_int(x1_in);
+            mp_int_t y1 = pb_obj_get_int(y1_in);
+            mp_int_t x2 = x2_in == mp_const_none ? source->image.width - 1 : pb_obj_get_int(x2_in);
+            mp_int_t y2 = y2_in == mp_const_none ? source->image.height - 1 : pb_obj_get_int(y2_in);
+            self->buffer = source->buffer;
+            self->is_display = false;
+            int width = x2 - x1 + 1;
+            int height = y2 - y1 + 1;
+            pbio_image_init_sub(&self->image, &source->image, x1, y1, width, height);
+        }
+    } else if (mp_obj_is_qstr(source_in) && MP_OBJ_QSTR_VALUE(source_in) == MP_QSTR__screen_) {
+        // Screen.
+        self->buffer = NULL;
+        self->is_display = true;
+        self->image = *pbdrv_display_get_image();
+    } else if (mp_obj_is_str(source_in)) {
+        // TODO: load from image "files".
+        mp_raise_ValueError(MP_ERROR_TEXT("'_screen_' is the only supported source string"));
```

Comment:
```suggestion
        mp_raise_NotImplementedError(MP_ERROR_TEXT("'_screen_' is the only supported source string"));
```

Assuming we intend to implement it.